[dynamo] Add size(), get_coordinate() support for DeviceMesh in dynamo

yoyoyocmu opened the pull request:
Summary: This fix is part of: https://github.com/pytorch/pytorch/issues/117670

Test Plan: Unit tetst and CI

Differential Revision: D52857348




cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

wconstab commented on torch/_dynamo/variables/distributed.py:
```diff
@@ -152,6 +152,34 @@ def var_getattr(self, tx, name: str) -> VariableTracker:
+    ) -> "VariableTracker":
+        if name == "size":
+            mesh_dim_var = None
+            if len(args) == 1:
```
Better to do more explicit/complete checking of args/kwargs- raise unsupported if getting unexpected combinations and try to support all the valid combinations (or if skipping some make a comment saying why)

wconstab reviewed (approved):
Please address my comment but overall lgtm thanks for the fix!

yoyoyocmu commented:
Latest version vs. initial version:
1. Resolve comments;
2. Add get_group() support;

wanchaol commented on torch/_dynamo/variables/distributed.py:
```diff
@@ -152,6 +152,42 @@ def var_getattr(self, tx, name: str) -> VariableTracker:
+        args: "List[VariableTracker]",
+        kwargs: "Dict[str, VariableTracker]",
+    ) -> "VariableTracker":
+        if name == "size":
```
no need to do in this PR, but I think the logics for each methods are very ad hoc, we should directly call the method instead of letting dynamo mimic the method calls, as it's not quite stable and can easily break if the actual method in eager changes.


As a follow up PR, I think we should do sth like:
```
args_as_value = [x.as_python_constant() for x in args]
kwargs_as_value = {k: v.as_python_constant() for k, v in kwargs.items()}
# call the method and wrap with ConstantVariable
...
```

Skylion007 replied:
At the very least, we should enable MyPy checking in the variables sub folder. It would catch a lot of bugs of undefined functions and methods.

wconstab replied:
args_as_value = [x.as_python_constant() for x in args]
kwargs_as_value = {k: v.as_python_constant() for k, v in kwargs.items()}
# call the method and wrap with ConstantVariable
...

Agree in this case for .size() its probably all constants.  But we can't blindly assume all args/kwargs should be converted to python constants, sometimes they would be tensorvariables/symints too.

And yes, would be better to call self.value.size() and wrap in constant.

otoh, are we sure it's constant?  shouldn't we be checking for dynamic size here?

yoyoyocmu replied:
Agree this will help make the code clean and generic, let me update it in next commit.

As far as I know, `.size()` is constant, and process groups are decided during initialization. I wonder if we have any cases that could make it dynamic?

anijain2305 replied:
@yanboliang Should inlining be easy with your recent tracing rules work?

wanchaol reviewed (approved):
lgtm, one comment we can follow up later.

anijain2305 commented on torch/_dynamo/variables/distributed.py:
```diff
@@ -152,6 +152,42 @@ def var_getattr(self, tx, name: str) -> VariableTracker:
+                return ConstantVariable.create(self.value.mesh.size(dim))
+        elif name == "get_coordinate":
+            coord = self.value.get_coordinate()
+            if coord is None:
+                return ConstantVariable.create(None)
+            else:
+                return ConstantVariable.create(coord)
```
Can be replaced by `return ConstantVariable.create(None)`

yoyoyocmu committed:
[dynamo] Add size(), get_coordinate() support for DeviceMesh in dynamo (#117710)

Summary:

This fix is part of: https://github.com/pytorch/pytorch/issues/117670

bypass-github-pytorch-ci-checks

Test Plan: Unit tetst and CI

Reviewed By: yanboliang, wanchaol

Differential Revision: D52857348
